docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,305 @@
+- `index.html` -> `/var/www/index.html`
+- `nginx.conf` -> `/var/nginx.conf`, `/etc/nginx.conf`
+
+A fundamental change compared to the current sync mechanism is that one source path may by sync'ed to multiple destinations in the future.
+
+### Assumptions
+
+Builders know where to put static files in the container.
+This information needs to be pulled from the builders and made available in Skaffold.
+Currently, we know how to do this for Docker artifacts (@r2d4 suggested this in #1166).
+Jib transforms some input files and includes others as static resources.
+The specific sync maps need to be detailed out but are should be straightforward.
+Whether the bazel builder supports this, is unclear at the moment.
+
+### Interface changes
+
+The builder implementations need to fulfill a new requirement:
+They need to provide a default sync map for changed or deleted files.
+For that, the `Builder` interface (pkg/skaffold/build/build.go) will receive the new method `SyncMap`.
+
+```go
+type Builder interface {
+   Labels() map[string]string
+
+   Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+   DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) ([]string, error)
+
+   // ADDED
+   SyncMap() (map[string][]string, error)
+}
+```
+
+`SyncMap` returns a map whose keys paths of source files for the container.
+These host paths will be relative to the workingdir of the respective artifact.
+The map values list all absolute remote destination paths where the input files are placed in the container.
+This has to be a list, because a single input file may be copied to several locations in the container.
+Note that this does not mean, that every file will be sync'ed.
+The intent to sync a file needs to be explicitly specified by user-provided sync rules (see config changes).
+In the future, some builders may specify default sync rules, but this is out of scope for this document.
+
+For example, given the above Dockerfile, it will return something like
+```go
+map[string][]string{
+  "static/page1.html": []string{"/var/www/page1.html"},
+  "static/assets/style.css": []string{"/var/www/assets/style.css"},
+  "index.html": []string{"/var/www/index.html"},
+  "nginx.conf": []string{"/var/nginx.conf", "/etc/nginx.conf"},
+}
+```
+
+Unsupported builders will return a not-supported error.
+
+**Note**: In the original version of this proposal, the signature of `GetDependencies` was changed. This is no longer the case
+
+### Config changes
+
+The `artifact.sync` config needs to support additional requirements:
+
+- Support existing use-cases for backwards compatibility, i.e.
+  - Glob patterns on the host path
+  - Transplant a folder structure, ~~or flatten the folder structure~~[this behavior was a bug and will be removed]
+- Offer a way to use destination inference.
+- Avoid ambiguity
+
+#### Problems with existing syntax
```

Comment:
This part provides a motivation for changing the syntax and might be worth moving to the top.

Reply:
moved it to the "Background" section